### DISCUSSION TITLE
Updating scoring algorithm task causes lag

### DISCUSSION THREAD

**Discussion by SollyBunny** [ppy/osu#26665] [Q&A]
### Type

Performance

### Bug description

Haven't played osu in a while, game started lagging midgame, I opened up notfications and saw this task running: when the progress went up my fps tanked

### Screenshots or videos

![image](https://github.com/ppy/osu/assets/76408142/03eb9cd9-fc44-4abc-bc35-77a145c13ff2)

### Version

2023.1231.0 Lazer

### Logs

[compressed-logs.zip](https://github.com/ppy/osu/files/13928699/compressed-logs.zip)

**Comment by bdach** [comment]
The processing is supposed to be taking place on a background thread. And logs claim that your cpu has "12 cores". So not really sure why this is happening.

Can you ctrl-f11 twice and attach a screenshot of that?

**Comment by peppy** [comment]
Sounds like realm might be blocking, if something on the main thread is also accessing realm. Hard to say what when the screenshot is cropped to show none of the game.

**Comment by SollyBunny** [comment]
> Sounds like realm might be blocking, if something on the main thread is also accessing realm. Hard to say what when the screenshot is cropped to show none of the game.

This happened in game, on the menu, and after a restart until the job completed. I don't really know how to un-update the scoring so I can replicate the issue

**Comment by peppy** [comment]
I've attempted to reproduce this but can't. There's a bit of overhead but it's not crippling. Worst case scenario of a 100mb+ realm database in single threaded mode:

![CleanShot 2024-01-22 at 08 40 44](https://github.com/ppy/osu/assets/191335/8909bf14-67b2-441d-9785-cbe46c8b721f)

**Comment by bdach** [comment]
I can reproduce this better with the new rank conversion thing in https://github.com/ppy/osu/pull/26630.

In profiling:

![1706015503](https://github.com/ppy/osu/assets/20418176/153bacef-30b0-4841-a397-0b3b76a419a3)

Not sure if there's something we can do... ~~Maybe take out a "stop all operations" for the duration of this?~~ that doesn't work since that blocks every thread including the calling one

**Comment by SollyBunny** [comment]
![image](https://github.com/ppy/osu/assets/76408142/d2aa8f49-b2cd-4ea4-b72f-3ef316751f8b)
still an issue

**Reply by SollyBunny** [reply]
if this can't be easily fixed can it be made lazy (as an option), ie only updates when it is required

**Reply by bdach** [reply]
> can it be made lazy (as an option), ie only updates when it is required

that is already the case.

**Reply by SollyBunny** [reply]
> > can it be made lazy (as an option), ie only updates when it is required
> 
> that is already the case.

on launching isn't really "when needed", i was more thinking only updating a single beatmap when you're hovering over it or playing it

**Comment by peppy** [comment]
If anyone reproduces this, please take a video of it happening with frame graphs visible (ctrl+f11 twice) and also the notification showing process of the migration. about 30 seconds of this should be plenty.
